DOC-224 (`hotfix`): Quickstart K8s Enterprise (#757)
localstack/localstack-docs: src/content/docs/aws/quickstart-library/k8s-quickstart.md

- [`kubectl`](https://kubernetes.io/docs/reference/kubectl/)
- [`jq`](https://jqlang.github.io/jq/download/)
- [`k9s`](https://k9scli.io/) (optional, for visual cluster monitoring)

## Step by Step

### Step 1: Clone the sample repository

```bash
git clone https://github.com/localstack-samples/localstack-k8s-demo.git
cd localstack-k8s-demo
```

The repository contains:

- `main.tf`: Terraform configuration that provisions an RDS MySQL database and a Lambda function on LocalStack
- `lambda-src/`: Python Lambda function source code with `pymysql` as a dependency
- `localstack-instance.yml`: Custom resource definition for the LocalStack deployment
- `scripts/`: Helper scripts for managing the auth token secret and port forwarding
- `Makefile`: Convenience targets for the full workflow

### Step 2: Create the Kubernetes cluster

If you want to monitor the cluster visually, open a separate terminal and run `k9s`. The interface starts empty but populates as pods come up.

Create a local Kubernetes cluster using `kind`:

```bash
kind create cluster --name ls-k8s-demo
```

Verify the cluster is running:

```bash
kubectl cluster-info
```

> ### Step 3: Deploy the LocalStack Operator

Reviewer: LocalStack operator is not available by default when someone starts a trial. We need to either:
1. Inform the customer about this and have them contact support or sales to get this added to their trial
2. Keep the getting started focussed on the helm deployment which shouldn't have any restrictions
3. make the operator and executor available by default on the trial

Note: Our kubernetes deployment is now much stronger that when we decided to take it off the trial so my recommendation would be #3

Author: Added note to address this, but I also think 3 is the best option here.

```bash
kubectl apply -f https://github.com/localstack/localstack-operator/releases/latest/download/controller.yaml
```

Wait for the operator pod to reach a `Running` state:

```bash
kubectl get pods -n localstack-operator-system
```

```
NAME                                                      READY   STATUS    RESTARTS   AGE
localstack-operator-controller-manager-78dcf78855-xxxxx   1/1     Running   0          30s
```

### Step 4: Deploy LocalStack into the cluster

Create a namespace and a secret containing your Auth Token:

```bash
kubectl create namespace workspace

kubectl create secret -n workspace generic localstack-auth-token \
    --from-literal=LOCALSTACK_AUTH_TOKEN=$LOCALSTACK_AUTH_TOKEN
```

Deploy the LocalStack instance:

```bash
kubectl apply --server-side -f ./localstack-instance.yml
```

Wait for the LocalStack pod to be ready (this may take a minute or two while the image is pulled):

```bash
kubectl get pods -n workspace -w
```

Proceed once the pod shows `1/1 Running`.

> ### Step 5: Set up port forwarding

Reviewer: I would add a section here to validate that the web app can connect to the localstack container as well. the webapp is an easy way to verify that your environment is up and running

Author: added this in latest commit

---

> title: "Quickstart: LocalStack on Kubernetes"

Reviewer 2: ```suggestion
title: "Deploy LocalStack on Kubernetes"
```

description: Deploy LocalStack into a Kubernetes cluster and run a sample Lambda + RDS application in under 5 minutes.
template: doc
sidebar:

>   order: 1

Reviewer 2: ```suggestion
  order: 3
```

---

## Introduction

> This quickstart gets LocalStack running in a local Kubernetes cluster in about 5 minutes. You will deploy a sample application consisting of a Lambda function that queries an RDS MySQL database, with both services running as pods in your cluster managed by LocalStack's Kubernetes executor.

Reviewer 2: ```suggestion
This quickstart spins up LocalStack in a local Kubernetes cluster and deploys a sample application in 5 minutes. You'll run an AWS Lambda function that queries an RDS MySQL database, with both services executing as pods managed by LocalStack.
```

:::

```bash
kubectl apply -f https://github.com/localstack/localstack-operator/releases/latest/download/controller.yaml
```

Wait for the operator pod to reach a `Running` state:

```bash
kubectl get pods -n localstack-operator-system
```

```
NAME                                                      READY   STATUS    RESTARTS   AGE
localstack-operator-controller-manager-78dcf78855-xxxxx   1/1     Running   0          30s
```

### Step 4: Deploy LocalStack into the cluster

Create a namespace and a secret containing your Auth Token:

```bash
kubectl create namespace workspace

kubectl create secret -n workspace generic localstack-auth-token \
    --from-literal=LOCALSTACK_AUTH_TOKEN=$LOCALSTACK_AUTH_TOKEN
```

Deploy the LocalStack instance:

```bash
kubectl apply --server-side -f ./localstack-instance.yml
```

Wait for the LocalStack pod to be ready (this may take a minute or two while the image is pulled):

```bash
kubectl get pods -n workspace -w
```

Proceed once the pod shows `1/1 Running`.

### Step 5: Set up port forwarding

> Forward port 4566 so you can run AWS commands against LocalStack from your local machine:

Reviewer 2: ```suggestion
Forward `port 4566` so you can run AWS commands against LocalStack from your local machine:
```

:::

### Step 7: Invoke the Lambda function

```bash
awslocal lambda invoke \
    --function-name myfunction \
    --payload '{}' /dev/stdout | jq .
```

The first invocation takes about 30 seconds as the Lambda pod starts up. You should see output like:

```json
{
  "results": [
    [1, "test"],
    [2, "another"]
  ]
}
```

## Validation

Confirm that all three pods are running in the `workspace` namespace:

```bash
kubectl get pods -n workspace
```

```
NAME                          READY   STATUS    AGE
lambda-myfunction-xxxxx       1/1     Running   36s
localstack-env-1-xxxxx        1/1     Running   11m
ls-mysql-xxxxx                1/1     Running   4m
```

You should see the LocalStack pod (`localstack-*`), the MySQL database pod (`ls-mysql-*`), and the Lambda function pod (`lambda-myfunction-*`) all running.

## Cleanup

To tear down all resources:

```bash
tflocal apply -destroy -auto-approve
kubectl delete -f ./localstack-instance.yml
kubectl delete secret -n workspace localstack-auth-token
```

> ## Troubleshooting
> 
> **LocalStack pod is stuck in `Pending` or `ImagePullBackOff`**
> Verify that your Auth Token secret was created correctly and that your cluster nodes can pull from the LocalStack registry. Check pod events with `kubectl describe pod -n workspace <pod-name>`.
> 
> **`awslocal sts get-caller-identity` times out**
> Confirm that port forwarding is still running in a separate terminal. If it dropped, restart it with `kubectl port-forward -n workspace svc/localstack-env-1 4566`.
> 
> **Lambda invocation returns an error after the first call**
> The first invocation takes up to 30 seconds for the Lambda pod to start. Wait and retry.
> 
> **Terraform apply fails with a connection error**
> Ensure port forwarding is active before running `tflocal apply`. LocalStack must be accessible on `localhost:4566`.
> 
> **MySQL pod does not start**
> Check cluster resource availability. The MySQL pod requires sufficient CPU and memory. Run `kubectl describe pod -n workspace <ls-mysql-pod-name>` to inspect scheduling events.
> 

Reviewer 2: ```suggestion
## Troubleshooting

### LocalStack pod is stuck in `Pending` or `ImagePullBackOff`
Verify that your Auth Token secret was created correctly and that your cluster nodes can pull from the LocalStack registry. Check pod events with `kubectl describe pod -n workspace <pod-name>`.

### `awslocal sts get-caller-identity` times out
Confirm that port forwarding is still running in a separate terminal. If it dropped, restart it with `kubectl port-forward -n workspace svc/localstack-env-1 4566`.

### Lambda invocation returns an error after the first call
The first invocation takes up to 30 seconds for the Lambda pod to start. Wait and retry.

### Terraform apply fails with a connection error
Ensure port forwarding is active before running `tflocal apply`. LocalStack must be accessible on `localhost:4566`.

### MySQL pod does not start
Check cluster resource availability. The MySQL pod requires sufficient CPU and memory. Run `kubectl describe pod -n workspace <ls-mysql-pod-name>` to inspect scheduling events.

```
